### PR TITLE
Draggable objects

### PR DESCRIPTION
**Pull request by denysLavrienov**
This PR makes files be draggable and movable to folders.
image

### CODE REVIEW

File: app/component/file-list-item.js [hunk 1]
```diff
@@ -150,10 +151,11 @@ export class FileListItem extends Component {
...
 
   /**
-   * Adds event handlers to `dragenter`, `dragover` and `dragleave` events.
-   * @param {Function} handler
+   * Adds event handlers to `dragenter`, `dragover` and `dragleave` and `drop` events.
+   * @param {Function} dropHandler
+   * @param {Function} moveHandler
    */
-  addDragEventListeners(handler) {
+  addDragEventListeners(dropHandler, moveHandler) {
```
Comment (dkalinin) [inline app/component/file-list-item.js]: Please fix.

File: app/component/file-list.js [hunk 2]
```diff
@@ -115,7 +117,16 @@ export class FileList extends StatefulComponent {
...
       this._dragEventHandler.drop(event, this._fileList);
       if (event.dataTransfer.files.length > 0) {
-        this.store.dispatch('uploadFiles', Array.from(event.dataTransfer.files));
+        this.store.dispatch('uploadFiles', Array.from(event.dataTransfer.files),
+            this.store.state.currentFolder.id)
+            .catch((error) => {
+              alert(error.message);
+            });
+      }
+      const fileData = event.dataTransfer.getData('file');
```
Comment (dkalinin) [inline app/component/file-list.js]: Please create a constant for 'file'.

File: app/service/actions.js [hunk 3]
```diff
@@ -17,26 +17,25 @@ export const actions = {
   uploadFiles: (store) => (files, folderId) => {
     store.commit('setFilesLoading', true);
     store.serviceFactory.getFileServiceInstance().uploadFiles(files, folderId).then(() => {
-      store.commit('addFiles', files, folderId);
+      // store.commit('addFiles', files, folderId);
```
Comment (dkalinin) [inline app/service/actions.js]: Please remove addFiles mutation.

File: app/service/actions.js [hunk 4]
```diff
@@ -63,4 +62,11 @@ export const actions = {
           store.commit('setCurrentFolder', folder);
         });
   },
+  moveFile: (store) => (fileId, folderId) => {
+    store.serviceFactory.getFileServiceInstance().moveFile(fileId, folderId)
+        .then(() => {
+          store.commit('moveFile', fileId, folderId);
```
Comment (dkalinin) [inline app/service/actions.js]: Let's remove this mutation.

File: app/component/file-list-item.js [hunk 5]
```diff
@@ -171,10 +188,19 @@ export class FileListItem extends Component {
       this._listItem.addEventListener('drop', (event) => {
         this._dragEventHandler.drop(event, this._listItem);
```
Comment (dkalinin) [inline app/component/file-list-item.js]: Please reduce duplicated code.

File: app/component/file-list-item.js [hunk 6]
```diff
@@ -24,6 +24,7 @@ export class FileListItem extends Component {
     this._dragEventHandler = new DragEventHandler();
     this._isFileRenaming = false;
     this._initialRender();
+    this._DATA_FORMAT = 'file';
```
Comment (dkalinin) [inline app/component/file-list-item.js]: This should be a standalone constant.

File: app/component/file-list-item.js [hunk 7]
```diff
@@ -171,10 +188,19 @@ export class FileListItem extends Component {
       this._listItem.addEventListener('drop', (event) => {
         this._dragEventHandler.drop(event, this._listItem);
         if (event.dataTransfer.files.length > 0) {
-          handler(Array.from(event.dataTransfer.files), this._file);
+          this._dropHandler(Array.from(event.dataTransfer.files), this._id);
+        }
+        const fileData = event.dataTransfer.getData(this._DATA_FORMAT);
+        if (!!fileData) {
```
Comment (dkalinin) [inline app/component/file-list-item.js]: !! is redundant.

File: app/service/actions.js [hunk 8]
```diff
@@ -16,27 +15,26 @@ export const actions = {
...
   uploadFiles: (store) => (files, folderId) => {
     store.commit('setFilesLoading', true);
-    store.serviceFactory.getFileServiceInstance().uploadFiles(files, folderId).then(() => {
-      store.commit('addFiles', files, folderId);
+    return store.serviceFactory.getFileServiceInstance().uploadFiles(files, folderId).then(() => {
     }).catch((error) => {
       console.error(error.message);
       alert(error.message);
+    }).finally(() => {
+      store.dispatch('fetchFiles');
```
Comment (dkalinin) [inline app/service/actions.js]: Should be done in then.

File: app/service/actions.js [hunk 9]
```diff
@@ -16,27 +15,26 @@ export const actions = {
...
-            .catch((error) => {
-              console.error(error.message);
-              alert(error.message);
-            })
-            .finally(() => {
-              store.commit('setFilesLoading', false);
-            }));
+    return store.serviceFactory.getFileServiceInstance().getFiles(store.state.currentFolder)
+        .then(() => {
+          store.commit('setFiles', store.state.files);
```
Comment (dkalinin) [inline app/service/actions.js]: Let's use files from backend response.

File: app/service/actions.js [hunk 10]
```diff
@@ -63,4 +63,10 @@ export const actions = {
           store.commit('setCurrentFolder', folder);
         });
   },
+  moveFile: (store) => (fileId, folderId) => {
+    store.serviceFactory.getFileServiceInstance().moveFile(fileId, folderId)
+        .then(() => {
+          store.dispatch('fetchFiles');
+        });
```
Comment (dkalinin) [inline app/service/actions.js]: Please handle errors.

File: app/service/file-service.js [hunk 11]
```diff
@@ -48,13 +48,13 @@ export class FileService {
...
+   * @param {string} fileID
+   * @param {string} folderID
    * @return {Promise}
    */
-  moveFile(file, folder) {
-    return this._requestService.put('/folder/' + file.id + '/move',
-        JSON.stringify(folder));
+  moveFile(fileID, folderID) {
+    return this._requestService.put('/folder/' + fileID + '/move',
+        folderID);
```
Comment (dkalinin) [inline app/service/file-service.js]: Switch folderId and fileId.

File: app/component/drag-event-handler.js [hunk 12]
```diff
@@ -45,4 +45,18 @@ export class DragEventHandler {
     this._enterCount--;
     node.classList.remove('drag-entered');
   }
+
+  handleDragEvents(node) {
+    node.addEventListener('dragenter', (event) => {
+      this.addEnterClass(event, node);
```
Comment (Oleh-Stefanovych) [inline app/component/drag-event-handler.js]: Let;s have separate method for `event.preventDefault();  event.stopPropagation();`

File: app/component/file-list-item.js [hunk 13]
```diff
@@ -152,7 +152,7 @@ export class FileListItem extends Component {
   }
 
   /**
-   * Adds drop handler for file list item if it is directory.
+   * Adds handleDropEvent handler for file list item if it is directory.
    * @param {Function} handler
    */
   addDropHandler(handler) {
```
Comment (dkalinin) [inline app/component/file-list-item.js]: Let's have more descriptive name for these handlers:
1. Is that file from OS was dropped.
2. Is that file from app was dropped.

File: app/component/file-list.js [hunk 14]
```diff
@@ -50,9 +50,9 @@ export class FileList extends StatefulComponent {
     this._listItems.className = 'list-unstyled list-items';
     this._fileList.appendChild(this._listItems);
     this.parentNode.appendChild(this._fileList);
-    this.store.dispatch('fetchFiles');
     this._showMessage('Loading...');
-    this._addEventListener();
+    this.store.dispatch('fetchFiles');
```
Comment (dkalinin) [inline app/component/file-list.js]: Render should not dispatch actions.

File: app/component/file-list.js [hunk 15]
```diff
@@ -62,9 +62,16 @@ export class FileList extends StatefulComponent {
    */
   _showMessage(text) {
     if (this.store.state.filesLoading) {
-      const span = document.createElement('span');
-      this._listItems.prepend(span);
-      span.textContent = text;
+      const loadingBlock = document.createElement('div');
+      loadingBlock.className = 'loading-block';
+      this._listItems.appendChild(loadingBlock);
```
Comment (dkalinin) [inline app/component/file-list.js]: Loading block should be out of `<ul>`.

File: app/component/file-list-item.js [hunk 16]
```diff
@@ -155,15 +155,15 @@ export class FileListItem extends Component {
    * Adds handleDropEvent handler for file list item if it is directory.
    * @param {Function} handler
    */
-  addDropHandler(handler) {
+  addFileFromOSDropHandler(handler) {
     this._dropHandler = handler;
```
Comment (dkalinin) [inline app/component/file-list-item.js]: Please rename according to method name.

File: app/component/file-list-item.js [hunk 17]
```diff
@@ -155,15 +155,15 @@ export class FileListItem extends Component {
...
   }
 
   /**
    * Adds move handler for file list item if it is directory.
    * @param {Function} handler
    */
-  addMoveHandler(handler) {
+  addFileFromListDropHandler(handler) {
     this._moveHandler = handler;
```
Comment (dkalinin) [inline app/component/file-list-item.js]: Please rename according to method name.

File: app/service/actions.js [hunk 18]
```diff
@@ -65,6 +65,7 @@ export const actions = {
   moveFile: (store) => (fileId, folderId) => {
     store.serviceFactory.getFileServiceInstance().moveFile(fileId, folderId)
         .then(() => {
+          store.commit('setFilesLoading', true);
```
Comment (dkalinin) [inline app/service/actions.js]: This should be set in fetchFiles action. Check other actions too.

**Pull request merged**